Haskell GHC ecosystem enumeration broken

**Describe the bug**
Enumeration of the Haskell GHC ecosystem is no longer working correctly

**To Reproduce**
Steps to reproduce the behaviour:
1. `make lib-tests` @ [35f782f](https://github.com/google/osv.dev/commit/35f782ff4cc000ad4aef65add2feb672dfaa155b)
2. observe `osv.ecosystems.haskell_test.GHCEcosystemTest.test_next_version` fails

**Expected behaviour**
Test does not fail

**Screenshots**
```
$ make lib-tests 
./run_tests.sh
+ export PIPENV_IGNORE_VIRTUALENVS=1
+ PIPENV_IGNORE_VIRTUALENVS=1
+ python3 -m pipenv sync
Installing dependencies from Pipfile.lock (9c09ee)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
All dependencies are now up-to-date!
+ python3 -m pipenv run python -m unittest osv.bug_test
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
+ python3 -m pipenv run python -m unittest osv.purl_helpers_test
......
----------------------------------------------------------------------
Ran 6 tests in 0.000s

OK
+ python3 -m pipenv run python -m unittest osv.request_helper_test
.
----------------------------------------------------------------------
Ran 1 test in 0.001s

OK
+ python3 -m pipenv run python -m unittest osv.semver_index_test
...
----------------------------------------------------------------------
Ran 3 tests in 0.001s

OK
+ python3 -m pipenv run python -m unittest osv.impact_test
....
----------------------------------------------------------------------
Ran 4 tests in 12.224s

OK
+ python3 -m pipenv run python -m unittest discover osv/ecosystems/ '*_test.py' .
.....F.........................
======================================================================
FAIL: test_next_version (osv.ecosystems.haskell_test.GHCEcosystemTest.test_next_version)
Test next_version.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/apollock/gosst/osv/osv.dev/osv/ecosystems/haskell_test.py", line 44, in test_next_version
    self.assertEqual('7.2.1', ecosystem.next_version('GHC', '7.0.4'))
AssertionError: '7.2.1' != '9.2.3'
- 7.2.1
+ 9.2.3


----------------------------------------------------------------------
Ran 31 tests in 108.714s

FAILED (failures=1)
make: *** [Makefile:15: lib-tests] Error 1
```

**Additional context**
It would appear that the behaviour of the GitLab API has changed

https://github.com/google/osv.dev/blob/35f782ff4cc000ad4aef65add2feb672dfaa155b/osv/ecosystems/haskell.py#L71-L72

is now returning only 20 entries, whereas I _suspect_ the old behaviour of passing `?per_page=-1` was to return all of them (`?per_page=100` certainly returns 100, and 20 appears to the default page size). Implementing [pagination](https://docs.gitlab.com/ee/api/rest/#pagination) is probably the solution.

--- a/osv/ecosystems/haskell.py
+++ b/osv/ecosystems/haskell.py
@@ -68,7 +68,6 @@ class Hackage(Ecosystem):
 class GHC(Ecosystem):
   """Glasgow Haskell Compiler (GHC) ecosystem."""
 
-  # FIXME: https://github.com/google/osv.dev/issues/2367
   _API_PACKAGE_URL = ('https://gitlab.haskell.org'
                       '/api/v4/projects/3561/repository/tags?per_page=100')
   """

--- a/osv/ecosystems/haskell.py
+++ b/osv/ecosystems/haskell.py
@@ -70,7 +70,7 @@ class GHC(Ecosystem):
 
   # FIXME: https://github.com/google/osv.dev/issues/2367
   _API_PACKAGE_URL = ('https://gitlab.haskell.org'
-                      '/api/v4/projects/3561/repository/tags?per_page=-1')
+                      '/api/v4/projects/3561/repository/tags?per_page=100')
   """
   Historical versions do not have tags in the Git repo, so we hardcode the
   list.  See https://github.com/google/osv.dev/pull/1463 for discussion.
@@ -136,18 +136,25 @@ class GHC(Ecosystem):
 
     """
 
-    response = requests.get(self._API_PACKAGE_URL, timeout=config.timeout)
-    if response.status_code == 404:
-      raise EnumerateError('GHC tag list not found')
-    if response.status_code != 200:
-      raise RuntimeError(f'Failed to get GHC versions with: {response.text}')
+    # Versions come from tags from the GitLab API, which are paginated.
+    # https://docs.gitlab.com/ee/api/rest/index.html#pagination-link-header
+    url = self._API_PACKAGE_URL
+    versions = self.HISTORICAL_VERSIONS.copy()
+    while url is not None:
+      response = requests.get(url, timeout=config.timeout)
+      if response.status_code == 404:
+        raise EnumerateError(f'GHC tag list not found at {url}')
+      if response.status_code != 200:
+        raise RuntimeError(
+            f'Failed to get GHC versions from: {url} with: {response.text}')
 
-    response = response.json()
-    versions = self.HISTORICAL_VERSIONS + [
-        self.tag_to_version(x['name'])
-        for x in response
-        if self.tag_to_version(x['name'])
-    ]
+      versions.extend(self.tag_to_version(x['name']) for x in response.json())
+
+      if 'next' not in response.links:
+        break
+      url = response.links['next'].get('url')
+
+    versions = [v for v in versions if v is not None]
 
     self.sort_versions(versions)
     return self._get_affected_versions(versions, introduced, fixed,

--- a/osv/ecosystems/haskell.py
+++ b/osv/ecosystems/haskell.py
@@ -70,6 +70,8 @@ class GHC(Ecosystem):
 
   _API_PACKAGE_URL = ('https://gitlab.haskell.org'
                       '/api/v4/projects/3561/repository/tags?per_page=100')
+  # 100 is the maximum per_page size according to GitLab docs:
+  # https://docs.gitlab.com/ee/api/rest/index.html#offset-based-pagination
   """
   Historical versions do not have tags in the Git repo, so we hardcode the
   list.  See https://github.com/google/osv.dev/pull/1463 for discussion.

--- a/osv/ecosystems/haskell_test.py
+++ b/osv/ecosystems/haskell_test.py
@@ -41,8 +41,7 @@ class GHCEcosystemTest(unittest.TestCase):
     self.assertEqual('0.29', ecosystem.next_version('GHC', '0'))
     self.assertEqual('7.0.4', ecosystem.next_version('GHC', '7.0.4-rc1'))
     # 7.0.4 is the last of the hardcoded versions
-    # Disabled due to https://github.com/google/osv.dev/issues/2367
-    # self.assertEqual('7.2.1', ecosystem.next_version('GHC', '7.0.4'))
+    self.assertEqual('7.2.1', ecosystem.next_version('GHC', '7.0.4'))
 
     # The whole GHC ecosystem is versioned together.  Enumeration ignores
     # package/component name.  Therefore this should NOT raise:

